lib/ansible/modules/cloud/vmware/vmware_guest_disk.py

```diff
@@ -0,0 +1,621 @@
+
+        return scsi_ctl
+
+    @staticmethod
+    def create_scsi_disk(scsi_ctl_key, disk_index):
+        """
+        Function to create Virtual Device Spec for virtual disk
+        Args:
+            scsi_ctl_key: Unique SCSI Controller Key
+            disk_index: Disk unit number at which disk needs to be attached
+
+        Returns: Virtual Device Spec for virtual disk
+
+        """
+        disk_spec = vim.vm.device.VirtualDeviceSpec()
+        disk_spec.operation = vim.vm.device.VirtualDeviceSpec.Operation.add
+        disk_spec.fileOperation = vim.vm.device.VirtualDeviceSpec.FileOperation.create
+        disk_spec.device = vim.vm.device.VirtualDisk()
+        disk_spec.device.backing = vim.vm.device.VirtualDisk.FlatVer2BackingInfo()
+        disk_spec.device.backing.diskMode = 'persistent'
+        disk_spec.device.controllerKey = scsi_ctl_key
+        disk_spec.device.unitNumber = disk_index
+        return disk_spec
+
+    def reconfigure_vm(self, config_spec, device_type):
+        """
+        Function to reconfigure virtual machine after modifying device spec
+        Args:
+            config_spec: Config Spec
+            device_type: Type of device being modified
+
+        Returns: Boolean status 'changed' and actual task result
+
+        """
+        changed, results = (False, '')
+        try:
+            # Perform actual VM reconfiguration
+            task = self.vm.ReconfigVM_Task(spec=config_spec)
+            changed, results = wait_for_task(task)
+        except vim.fault.InvalidDeviceSpec as invalid_device_spec:
+            self.module.fail_json(msg="Failed to manage %s on given virtual machine due to invalid"
+                                      " device spec : %s" % (device_type, to_native(invalid_device_spec.msg)),
+                                  details="Please check ESXi server logs for more details.")
+        except vim.fault.RestrictedVersion as e:
+            self.module.fail_json(msg="Failed to reconfigure virtual machine due to"
+                                      " product versioning restrictions: %s" % to_native(e.msg))
+
+        return changed, results
+
+    def ensure_disks(self, vm_obj=None):
+        """
+        Function to manage internal state of virtual machine disks
+        Args:
+            vm_obj: Managed object of virtual machine
+
+        """
+        # Set vm object
+        self.vm = vm_obj
+        # Sanitize user input
+        disk_data = self.sanitize_disk_inputs()
+        # Create stateful information about SCSI devices
+        current_scsi_info = dict()
+        results = dict(changed=False, disk_data=None, disk_changes=dict())
+
+        # Deal with SCSI Controller
+        for device in vm_obj.config.hardware.device:
+            if isinstance(device, tuple(self.scsi_device_type.values())):
+                # Found SCSI device
+                if device.busNumber not in current_scsi_info:
+                    current_scsi_info[device.key] = dict(disks=dict())
```

Comment:
you check if `device.busNumber` isn't a key of the dict `current_scsi_info` but then you add the key `device.key`.
You should check for `device.busNumber` and then create the key `device.busNumber`. Or check for `device.key` and then create the key `device.key`.

Am I missing something ?